parsers/VN.py

```diff
@@ -1,216 +1,163 @@
+    result_list = TotalConsumptionList(logger)
 
+    for data in data_list:
         result_list.append(
-            {
-                "datetime": i_datetime.datetime,
-                "consumption": per_zone_load[zone_key][i],
-                "zoneKey": zone_key,
-                "source": "nldc.evn.vn",
-            }
-        )
-
-    if not len(result_list):
-        raise ParserException(
-            parser="VN.py",
-            message=f"No valid consumption data for requested day found.",
+            datetime=datetime.fromisoformat(data["thoiGian"]).replace(tzinfo=tz),
+            consumption=data[f"congSuat{ZONE_CODE[ZoneKey('VN')]}"],
+            zoneKey=ZoneKey("VN"),
+            source="nldc.evn.vn",
         )
+    return result_list.to_list()
 
-    return result_list
 
-
-def fetch_price(
-    zone_key: str,
+def fetch_live_price(
     session: Session = Session(),
-    target_datetime: datetime | None = None,
     logger: Logger = getLogger(__name__),
-) -> list[dict] | dict:
-    request_latest = target_datetime is None
+):
+    data_list = fetch_live_data("price", session)["giaBiens"]
 
-    if request_latest:
-        fetched_day, fetched_data = fetch_latest_data("price", session)
-    else:
-        # TODO API bug for past data:
-        # data for given date can only be retrieved until the current hour.
-        fetched_day, fetched_data = fetch_target_data("price", target_datetime, session)
-
-    # [0] = load of Northern Vietnam
-    # [1] = load of Central Vietnam
-    # [2] = load of Southern Vietnam
-    per_zone_price = {
-        "VN-N": fetched_data["data"][0],
-        "VN-C": fetched_data["data"][1],
-        "VN-S": fetched_data["data"][2]
-        # no total VN data
-    }
-
-    result_list = []
-
-    for i in range(len(per_zone_price[zone_key])):
-        # Convert i to datetime
-        i_datetime = data_index_to_valid_time(fetched_day, i)
-
-        # Data also goes ~2h into the future! Seems to be some kind of prediction, since these
-        # future values are changing until their valid time is reached.
-        if request_latest and i_datetime > arrow.now(tz=tz):
-            continue
-
-        # faulty data at given time?
-        if per_zone_price[zone_key][i] is None:
-            continue
+    result_list = PriceList(logger)
 
+    for data in data_list:
         result_list.append(
-            {
-                "datetime": i_datetime.datetime,
-                "price": per_zone_price[zone_key][i],
-                "currency": "VND",
-                "zoneKey": zone_key,
-                "source": "nldc.evn.vn",
-            }
+            datetime=datetime.fromisoformat(data["thoiGian"]).replace(tzinfo=tz),
+            price=data[f"giaBien{ZONE_CODE[ZoneKey('VN')]}"],
+            currency="VND",
+            zoneKey=ZoneKey("VN"),
+            source="nldc.evn.vn",
         )
+    return result_list.to_list()
+
+
+def fetch_historical_data(
+    target_datetime: datetime,
```

Comment:
Since target_datetime can be None this need to either be updated to allow None or it needs to be handled earlier so target_datetime is never None.